### PR TITLE
Bioenv 0 variance filter

### LINKED ISSUE #275: Bioenv fails if 0 variance columns are left after dropping rows with empty cells
**Bug Description**
If `bioenv` gets any 0 variance numerical columns, it errors. There is currently a safeguard in place to prevent 0 variance columns from ever reaching the stage where they become an issue; however, after this safeguard we drop all rows containing any empty cells. If any column becomes 0 variance after dropping these rows, we still get the error.

**Steps to reproduce the behavior**
1. Run the following commands
```
wget \
  -O "sample-metadata.tsv" \
  "https://data.qiime2.org/2020.2/tutorials/atacama-soils/sample_metadata.tsv"

qiime metadata distance-matrix \
  --m-metadata-file sample-metadata.tsv \
  --m-metadata-column elevation \
  --o-distance-matrix elevation-distance-matrix.qza

qiime diversity bioenv \
  --i-distance-matrix elevation-distance-matrix.qza \
  --m-metadata-file sample-metadata.tsv \
  --o-visualization bioenv-elevation.qzv \
  --verbose
```
2. Error

**Expected behavior**
Filter zero variance columns after removal of rows containing empty cells so this doesn't happen.

**Screenshots**
image

**Computation Environment**
- OS: Linux Mint 19.3 (forum user was using virtual box)
- QIIME 2 2020.2

**References**
1. [Forum](https://forum.qiime2.org/t/qiime-diversity-bioenv-error-when-running-through-atacama-soil-microbiome-tutorial/15272/7) post that prompted this.
2. Ancient [forum](https://forum.qiime2.org/t/diversity-bioenv-column-s-in-the-data-frame-could-not-be-scaled/58/3) post in regards to the same thing.
3. Ancient [PR](https://github.com/qiime2/q2-diversity/pull/72) discussed in that forum post that seems to have been superceded.

### PR DESCRIPTION
**Pull request by Oddant1**
closes #275

### CODE REVIEW

**Comment by andrewsanchez** [comment]
This seems fine to me if jumping between QIIME 2 objects and pandas DataFrames is a common/acceptable pattern.  Is that the case?  

However, should we consider extending `Metadata.filter_ids` to handle stuff like this?  Even so, I'm good with merging this now and handling opening another issue for that.

File: q2_diversity/_beta/_visualizer.py [hunk 1]
```diff
@@ -40,17 +40,25 @@ def bioenv(output_dir: str, distance_matrix: skbio.DistanceMatrix,
     metadata = metadata.filter_columns(column_type='numeric')
     non_numeric_cols = pre_filtered_cols - set(metadata.columns)
 
+    # Drop samples that have any missing values.
+    # TODO use Metadata API if this type of filtering is supported in the
+    # future.
+    # NOTE this whole caste to df to dropnans then caste back to metadata for
```
Comment (ChrisKeefe) [inline q2_diversity/_beta/_visualizer.py:46]:
```suggestion
    # NOTE this whole cast to df to dropnans then cast back to metadata for
```
`cast` > `caste`
:heart: Number One Spelling Jerk

**Comment by thermokarst** [comment]
> This seems fine to me if jumping between QIIME 2 objects and pandas DataFrames is a common/acceptable pattern. Is that the case?

It isn't ideal, but its not a showstopper, in my opinion. I would love to see a new filter added to `Metadata`, though, that would be a sweet addition.

File: q2_diversity/_beta/_visualizer.py [hunk 2]
```diff
@@ -40,17 +40,25 @@ def bioenv(output_dir: str, distance_matrix: skbio.DistanceMatrix,
...
+    # TODO use Metadata API if this type of filtering is supported in the
+    # future.
+    # NOTE this whole cast to df to dropnans then cast back to metadata for
+    # other filtering then go back to df again to use is gross. It doesn't look
+    # like metadata currently supports this type of filtering (as per the
+    # above TODO), but this sure makes me wish it did.
+    df = metadata.to_dataframe()
+    df = df.dropna(axis='index', how='any')
+
+    metadata = qiime2.Metadata(df)
```
Comment (andrewsanchez) [inline q2_diversity/_beta/_visualizer.py:53]: @thermokarst I just wanted to clean up the comments here before merging.  Shall I go ahead and tidy up this PR myself, or wait until Anthony is back in action?

```suggestion
    # Drop samples that have any missing values.
    # TODO use Metadata API if more filtering is supported in the future.
    df = metadata.to_dataframe()
    df = df.dropna(axis='index', how='any')
    metadata = qiime2.Metadata(df)
```

**Comment by andrewsanchez** [comment]
> > This seems fine to me if jumping between QIIME 2 objects and pandas DataFrames is a common/acceptable pattern. Is that the case?
> 
> It isn't ideal, but its not a showstopper, in my opinion. I would love to see a new filter added to `Metadata`, though, that would be a sweet addition.

Cool, then I'll go ahead and merge this then open a new ENH issue on the framework about adding a new filter.

File: q2_diversity/_beta/_visualizer.py (hunk 2 above)
Comment (thermokarst) [reply q2_diversity/_beta/_visualizer.py:53]: Thanks for checking! I checked in with @Oddant1 earlier this week - he said to go ahead and push up any edits necessary to close up his PRs. Merge away!

File: q2_diversity/_beta/_visualizer.py [hunk 3]
```diff
@@ -40,17 +40,19 @@ def bioenv(output_dir: str, distance_matrix: skbio.DistanceMatrix,
     metadata = metadata.filter_columns(column_type='numeric')
     non_numeric_cols = pre_filtered_cols - set(metadata.columns)
 
+    # Drop samples that have any missing values.
+    # TODO use Metadata API if more filtering is supported in the future.
+    df = metadata.to_dataframe()
+    df = df.dropna(axis='index', how='any')
```
Comment (andrewsanchez) [inline q2_diversity/_beta/_visualizer.py:46]:
```suggestion
    df = df.dropna()
```

**Pull request merged**